libuavcan/include/libuavcan/transport/media/can.hpp

```diff
@@ -0,0 +1,336 @@
+    static constexpr std::uint32_t FlagRTR   = 1U << 30;  ///< Remote transmission request
+    static constexpr std::uint32_t FlagERR   = 1U << 29;  ///< Error frame
+
+    static constexpr std::uint16_t MtuBytes = MtuBytesParam;
+
+    // TODO: allow additional meta-data like ts valid flag
+    // and "after rx overflow".
+    // This should be extensible.
+
+    static FrameDLC lengthToDlc(std::uint_fast8_t length)
+    {
+        /**
+         * Lookup table to map a CAN frame length to a DLC value
+         * that will accommodate the frame.
+         */
+        constexpr std::uint8_t length_to_dlc_lookup[] = {0,  1,  2,  3,  4,  5,  6,  7,  8,  9,  9,  9,  9,
+                                                         10, 10, 10, 10, 11, 11, 11, 11, 12, 12, 12, 12, 13,
+                                                         13, 13, 13, 13, 13, 13, 13, 14, 14, 14, 14, 14, 14,
+                                                         14, 14, 14, 14, 14, 14, 14, 14, 14, 14, 15, 15, 15,
+                                                         15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15};
+
+        static_assert(std::extent<decltype(length_to_dlc_lookup)>::value == 65,
+                      "length_to_dlc_lookup table is malformed.");
+        static_assert(std::extent<decltype(length_to_dlc_lookup)>::value > MtuBytesParam,
+                      "CAN MTU cannot exceed the length of the dlc lookup table.");
+
+        FrameDLC result;
+        if (length > MtuBytesParam)
+        {
+            result = FrameDLC(length_to_dlc_lookup[MtuBytesParam]);
+        }
+        else
+        {
+            // Because MtuBytesParam cannot exceed the length of the lookup table
+            // (see static assert) it is safe to use length as an index in the branch
+            // without further checks.
+            result = FrameDLC(length_to_dlc_lookup[length]);
+        }
+        return result;
+    }
+
+    static std::uint_fast8_t dlcToLength(FrameDLC dlc)
+    {
+        /**
+         * Lookup table to map a DLC value to the maximum data
+         * payload length supported for the DLC.
+         */
+        constexpr std::uint8_t dlc_to_length_lookup[] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 12, 16, 20, 24, 32, 48, 64};
+
+        static_assert(std::extent<decltype(dlc_to_length_lookup)>::value == 16,
+                      "dlc_to_length_lookup table is malformed.");
+
+        const auto dlc_value = static_cast<std::underlying_type<FrameDLC>::type>(dlc);
+        if (dlc_value < std::extent<decltype(dlc_to_length_lookup)>::value)
+        {
+            return dlc_to_length_lookup[dlc_value];
+        }
+        else
+        {
+            return 0;
+        }
+    }
+
+    std::uint32_t id;  ///< CAN ID with flags (above)
+    std::uint8_t  data[MtuBytesParam];
+
+private:
+    FrameDLC dlc;  ///< Data Length Code
+
+public:
+    Frame()
+        : id(0)
+        , dlc(FrameDLC::CodeForLength0)
```

Comment:
You are referring to the `data` datamember, correct? It's size is determined by the template so I'm unsure how I would do this.

Reply:
No, I am referring to this:

```c++
    std::uint32_t id = 0;  ///< CAN ID with flags (above)
    std::uint8_t  data[MtuBytesParam]{};  // Zero initialization

private:
    FrameDLC dlc_ = FrameDLC::FrameDLC::CodeForLength0;  ///< Data Length Code
```

> It's size is determined by the template so I'm unsure how I would do this.

This is not what I was referring to initially but see the sample above.

Reply:
Oh. Gah. I hate initializing things like that. For one thing, when you are in a debugger, you jump all over the class when stepping through the constructor which is a symptom of the other thing, you are doing initialization in several places in the code (i.e. you do some in the inline declarations and some in the constructor itself). I'd rather stick to initializer lists if you don't mind.

Reply:
The reason I have brought this up is because I am concerned that by continuing our reliance on member initializer lists we might be undermining correctness and maintainability of the code (which might be tightly correlated?). Having consulted with HIC++ I am now observing that the code is actually perfectly correct and I am at fault here for failing to recall the HIC++ rule 12.4.3 (shame on me) which says that non-static data member initializers are not to be used with members which are initialized differently by overloaded constructors.

Both HIC++ (rule 12.4.2) and C++CG (rule C.45, indirectly), however, say that non-static data member initializers shall be used instead of constructor initializer lists for data members that are initialized identically by all available overloaded constructors. Neither of the rules apply to this class so this can be ignored.

I am sorry for being unnecessarily pedantic here.

Reply:
Is there an open-source HIC++ checker available we could integrate with our build?

Reply:
I am not aware of any, but clang-tidy might be an adequate replacement.

Reply:
I added #226 to track this.